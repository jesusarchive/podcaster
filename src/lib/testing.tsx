import { render } from '@testing-library/react';
import React from 'react';
import { createMemoryRouter, RouterProvider } from 'react-router-dom';

/**
 * Testing utilities
 *
 * Centrailize testing utilities for unit tests indepedent of the testing library used.
 */

// Render a component with react-router-dom BrowserRouter wrapper
// export function renderWithRouter(ui, { route = '/' } = {}) {
//   window.history.pushState({}, 'Test page', route);

//   return {
//     ...render(ui, { wrapper: BrowserRouter })
//   };
// }

export function renderWithRouter(element, loader) {
  const routes = [
    {
      path: '/',
      element,
      loader
    }
  ];

  const router = createMemoryRouter(routes, { initialEntries: ['/'] });

  return {
    ...render(<RouterProvider router={router} />)
  };
}
