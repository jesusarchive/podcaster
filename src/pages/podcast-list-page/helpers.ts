import { PodcastType } from '@/services/podcasts';

export function filterPodcasts(podcasts: Array<PodcastType>, search: string) {
  return podcasts.filter((podcast) => {
    if (!search) {
      return true;
    }

    return (
      podcast['im:name'].label.toLowerCase().includes(search.toLowerCase()) ||
      podcast['im:artist'].label.toLowerCase().includes(search.toLowerCase())
    );
  });
}
