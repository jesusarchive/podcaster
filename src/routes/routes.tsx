import React from 'react';
import { createBrowserRouter, Outlet, ScrollRestoration } from 'react-router-dom';

import ErrorBoundary from '@/components/error-boundary';
import Header from '@/layouts/header';
import EpisodeDetailPage, { episodeDetailPageLoader } from '@/pages/episode-detail-page';
import NotFoundPage from '@/pages/not-found-page';
import PodcastDetailPage, { podcastDetailPageLoader } from '@/pages/podcast-detail-page';
import PodcastList, { podcastListPageLoader } from '@/pages/podcast-list-page';

export const routesConfig = [
  {
    path: '/',
    element: (
      <>
        <Header />
        <Outlet />
        <ScrollRestoration />
      </>
    ),
    children: [
      {
        index: true,
        element: <PodcastList />,
        loader: podcastListPageLoader,
        errorElement: <ErrorBoundary />
      },
      {
        path: 'podcast/:podcastId',
        element: <PodcastDetailPage />,
        loader: podcastDetailPageLoader,
        errorElement: <ErrorBoundary />
      },
      {
        path: 'podcast/:podcastId/episode/:episodeId',
        element: <EpisodeDetailPage />,
        loader: episodeDetailPageLoader,
        errorElement: <ErrorBoundary />
      },
      {
        path: '*',
        element: <NotFoundPage />
      }
    ]
  }
];

export default createBrowserRouter(routesConfig);
