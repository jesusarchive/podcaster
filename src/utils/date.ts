export function formatDate(date: string): string {
  return new Date(date).toLocaleDateString();
}

export const millisToHms = (millis: number): string => {
  const hours = Math.floor(millis / 3600000);
  const minutes = Math.floor((millis % 3600000) / 60000);
  const seconds = Math.floor(((millis % 3600000) % 60000) / 1000);

  const hoursStr = hours > 0 ? `${hours}:` : '';
  const minutesStr = minutes > 0 ? `${minutes < 10 ? `0${minutes}` : minutes}:` : '00:';
  const secondsStr = seconds > 0 ? `${seconds < 10 ? `0${seconds}` : seconds}` : '00';

  return `${hoursStr}${minutesStr}${secondsStr}`;
};

export function hmsToSeconds(time: string): number {
  const [seconds, minutes, hours = 0] = time
    .split(':')
    .map((el) => Number(el))
    .reverse();

  return hours * 3600 + minutes * 60 + seconds;
}
