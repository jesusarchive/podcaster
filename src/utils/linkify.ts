/**
 * Linkify text
 *
 * Replace URLs and emails in text with anchor tags.
 */
export function linkify(text: string): string {
  const urlRegex =
    /(https?:\/\/(?:www\.|(?!www))[a-zA-Z0-9][a-zA-Z0-9-]+[a-zA-Z0-9]\.[^\s]{2,}|www\.[a-zA-Z0-9][a-zA-Z0-9-]+[a-zA-Z0-9]\.[^\s]{2,}|https?:\/\/(?:www\.|(?!www))[a-zA-Z0-9]+\.[^\s]{2,}|www\.[a-zA-Z0-9]+\.[^\s]{2,})/gi;
  const emailRegex = /([\w.]+@[\w.]+\.[\w]+)/g;

  return text
    .replace(urlRegex, (url) => {
      let hyperlink = url;
      if (!hyperlink.match('^https?://')) {
        hyperlink = `http://${hyperlink}`;
      }

      return `<a href="${hyperlink}" target="_blank" rel="noopener noreferrer">${url}</a>`;
    })
    .replace(emailRegex, (email) => {
      return `<a href="mailto:${email}">${email}</a>`;
    });
}
