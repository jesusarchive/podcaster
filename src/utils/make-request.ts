/**
 * Make request
 *
 * Centralize api utility functions, independent of the library used. E.g. fetch, axios, etc.
 *
 * Fetch data from api and handle errors.
 */
export async function makeRequest(input: RequestInfo | URL, init?: RequestInit) {
  const response = await fetch(input, init);

  if (!response.ok) {
    throw new Error(response.statusText);
  }

  const data = await response.json();

  if (data && data.error) {
    throw new Error(data.error);
  }

  return data;
}
