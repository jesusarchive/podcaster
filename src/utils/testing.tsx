/**
 * Testing utilities
 *
 * Centrailize testing utilities for unit tests indepedent of the testing library used.
 */

// Render a component with react-router-dom BrowserRouter wrapper
// export function renderWithRouter(ui, { route = '/' } = {}) {
//   window.history.pushState({}, 'Test page', route);

//   return {
//     ...render(ui, { wrapper: BrowserRouter })
//   };
// }

// export function renderWithRouter(element, loader) {
//   const routes = [
//     {
//       path: '/',
//       element,
//       loader
//     }
//   ];

//   const router = createMemoryRouter(routes, { initialEntries: ['/'] });

//   return {
//     ...render(<RouterProvider router={router} />)
//   };
// }

export function getWindow(initialUrl: string): Window {
  window.history.replaceState(null, '', initialUrl);

  return window as unknown as Window;
}

export function createDeferred() {
  let resolve: (val?: any) => Promise<void>;
  let reject: (error?: Error) => Promise<void>;
  const promise = new Promise((res, rej) => {
    resolve = async (val: any) => {
      console.warn('RESOLVE', val);
      res(val);
      try {
        await promise;
      } catch (e) {}
    };
    reject = async (error?: Error) => {
      rej(error);
      try {
        await promise;
      } catch (e) {}
    };
  });

  return {
    promise,
    resolve,
    reject
  };
}
