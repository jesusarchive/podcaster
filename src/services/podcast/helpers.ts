import { isDateOlderThanDays } from '@/utils/date';

import { LOCAL_STORAGE_CLEANUP_DAYS, LOCAL_STORAGE_CREATED_AT_KEY } from './constants';

// handle local storage cleanup if it was created more than a day ago
function handleLocalStorageCleanup() {
  const createdAt = localStorage.getItem(LOCAL_STORAGE_CREATED_AT_KEY);

  if (!createdAt) {
    localStorage.setItem(LOCAL_STORAGE_CREATED_AT_KEY, String(new Date()));
  } else {
    const parsedCreatedAt = new Date(createdAt);

    if (isDateOlderThanDays(parsedCreatedAt, LOCAL_STORAGE_CLEANUP_DAYS)) {
      localStorage.clear();
      localStorage.setItem(LOCAL_STORAGE_CREATED_AT_KEY, String(new Date()));
      window.location.href = '/';
    }
  }
}

// get parsed local storage data and handle cleanup
export function getLocalStorageData<T>(key: string): T | undefined {
  handleLocalStorageCleanup();
  const rawLocalStorageData = localStorage.getItem(key);

  return rawLocalStorageData && JSON.parse(rawLocalStorageData);
}
