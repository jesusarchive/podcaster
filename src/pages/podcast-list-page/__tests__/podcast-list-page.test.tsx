import { prettyDOM, render, screen, waitFor } from '@testing-library/react';
import React from 'react';
import {
  createBrowserRouter,
  createRoutesFromElements,
  Outlet,
  Route,
  RouterProvider,
  useLoaderData,
  useLocation
} from 'react-router-dom';

import { routesConfig } from '@/routes';
import topPodcastsResponseMock from '@/services/podcast/__mocks__/top-podcasts-response-mock.json';
import { createDeferred, getWindow } from '@/utils/testing';

import { filterPodcasts } from '../helpers';
import PodcastListPage from '../podcast-list-page';

describe('<PodcastListPage />', () => {
  beforeAll(() => {
    (global.fetch as jest.Mock) = jest.fn(() =>
      Promise.resolve({
        json: () => Promise.resolve({})
      })
    );
  });

  it('renders podcast list page', async () => {
    const router = createBrowserRouter(routesConfig, {
      window: getWindow('/'),
      hydrationData: {
        loaderData: {
          podcastPromise: []
        }
      }
    });

    const { container } = render(<RouterProvider router={router} />);

    expect(container).toBeTruthy();
  });

  it('should filter podcasts', () => {
    const podcasts = topPodcastsResponseMock.feed.entry;
    const result = filterPodcasts(podcasts, 'counterclock');
    expect(result).toEqual([podcasts[0]]);
  });

  it('should return all podcasts', () => {
    const podcasts = topPodcastsResponseMock.feed.entry;
    const result = filterPodcasts(podcasts, '');
    expect(result).toEqual(podcasts);
  });

  it('renders', () => {
    window.__staticRouterHydrationData = {
      loaderData: { hello: 'world' }
    };

    const router = createBrowserRouter(createRoutesFromElements(<Route path="/" element={<PodcastListPage />} />));

    render(<RouterProvider router={router} />);
    expect(router).toBeTruthy();
  });

  it('new podcast list page test', async () => {
    const dfd = createDeferred();

    window.__staticRouterHydrationData = {
      loaderData: {
        '0': {
          podcastPromise: () => dfd.promise
        }
      }
    };
    const router = createBrowserRouter(
      createRoutesFromElements(<Route path="/" element={<PodcastListPage />}></Route>),
      {
        window: getWindow('/')
      }
    );
    const { container } = render(<RouterProvider router={router} />);
    await dfd.resolve(topPodcastsResponseMock.feed.entry);
    // await waitFor(() => screen.getByText('TEST PODCASTS'));

    expect(container).toBeTruthy();
    expect(container).toMatchSnapshot();
  });

  it('new podcast list page test2', async () => {
    const dfd = createDeferred();

    window.__staticRouterHydrationData = {
      loaderData: {
        '0': {
          test: 'test',
          test2: () => dfd.promise
          // podcastPromise: topPodcastsResponseMock.feed.entry
        }
      }
    };
    const router = createBrowserRouter(
      createRoutesFromElements(<Route path="/" element={<PodcastListPage />}></Route>),
      {
        window: getWindow('/')
      }
    );
    const { container } = render(<RouterProvider router={router} />);
    await dfd.resolve('test2');
    // await dfd.resolve(topPodcastsResponseMock.feed.entry);
    await waitFor(async () => await dfd.resolve('test2'));

    expect(container).toBeTruthy();
    expect(container).toMatchSnapshot();
  });

  it('new podcast list page test3', async () => {
    const dfd = createDeferred();

    window.__staticRouterHydrationData = {
      loaderData: {
        '0': {
          test: 'test',
          test2: () => dfd.promise
          // podcastPromise: topPodcastsResponseMock.feed.entry
        }
      }
    };
    const router = createBrowserRouter(
      createRoutesFromElements(<Route path="/" element={<PodcastListPage />}></Route>),
      {
        window: getWindow('/')
      }
    );
    const { container } = render(<RouterProvider router={router} />);
    await dfd.resolve('test2');
    // await dfd.resolve(topPodcastsResponseMock.feed.entry);
    await waitFor(async () => await dfd.resolve('test2'));

    expect(container).toBeTruthy();
    expect(container).toMatchSnapshot();
  });

  // it('new new podcast list page test', async () => {
  //   const dfd = createDeferred();

  //   const router = createBrowserRouter([
  //     {
  //       path: '/',
  //       element: <PodcastListPage />,
  //       loader: () => ({ podcastPromise: () => dfd.promise })
  //     }
  //   ]);

  //   const { container } = render(<RouterProvider router={router} />);
  //   await dfd.resolve(topPodcastsResponseMock.feed.entry);
  //   // await waitFor(() => screen.getByText('TEST PODCASTS'));
  //   expect(container).toBeTruthy();
  //   expect(container).toMatchSnapshot();
  // });

  it('renders fallbackElement within router contexts', async () => {
    const fooDefer = createDeferred();
    const router = createBrowserRouter(
      createRoutesFromElements(
        <Route path="/" element={<Outlet />}>
          <Route path="foo" loader={() => fooDefer.promise} element={<Foo />} />
        </Route>
      ),
      { window: getWindow('/foo') }
    );
    const { container } = render(<RouterProvider router={router} fallbackElement={<FallbackElement />} />);

    function FallbackElement() {
      const location = useLocation();

      return <p>Loading{location.pathname}</p>;
    }

    function Foo() {
      const data = useLoaderData() as { message: string };

      return <h1>Foo:{data?.message}</h1>;
    }

    function getHtml(container: HTMLElement) {
      return prettyDOM(container, undefined, {
        highlight: false,
        theme: {
          comment: null,
          content: null,
          prop: null,
          tag: null,
          value: null
        }
      });
    }

    expect(getHtml(container)).toMatchSnapshot();

    fooDefer.resolve({ message: 'From Foo Loader' });
    await waitFor(() => screen.getByText('Foo:From Foo Loader'));
    expect(getHtml(container)).toMatchSnapshot();
  });
});
