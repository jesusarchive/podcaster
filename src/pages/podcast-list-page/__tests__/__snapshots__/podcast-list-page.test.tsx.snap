// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PodcastListPage /> new podcast list page test 1`] = `
<div>
  <main
    class="podcast-list-page"
  >
    <span>
      idle
    </span>
    <span>
      {}
    </span>
    <span>
      "object"
    </span>
    <span>
      No podcasts found. 1
      {}
    </span>
  </main>
</div>
`;

exports[`<PodcastListPage /> new podcast list page test2 1`] = `
<div>
  <main
    class="podcast-list-page"
  >
    <span>
      idle
    </span>
    <span>
      {"test":"test"}
    </span>
    <span>
      "object"
    </span>
    <span>
      No podcasts found. 1
      {"test":"test"}
    </span>
  </main>
</div>
`;

exports[`<PodcastListPage /> new podcast list page test3 1`] = `
<div>
  <main
    class="podcast-list-page"
  >
    <span>
      idle
    </span>
    <span>
      {"test":"test"}
    </span>
    <span>
      "object"
    </span>
    <span>
      No podcasts found. 1
      {"test":"test"}
    </span>
  </main>
</div>
`;

exports[`<PodcastListPage /> renders fallbackElement within router contexts 1`] = `
"<div>
  <h1>
    Foo:
  </h1>
</div>"
`;
